File: src/main/java/com/roughlyunderscore/enchantsapi/registry/RegistrableCondition.java
package com.roughlyunderscore.enchantsapi.registry;

import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.event.Event;
import org.jetbrains.annotations.NotNull;

/**
 * This interface is used to register conditions. Implement this interface to create a new condition and
 * override each and every method. By default, each method has a default implementation that throws an exception.
 * @see UEAPIRegistry
 * @since 2.2
 * @author Roughly_
 */
public interface RegistrableCondition extends Registrable {
  /**
   * Evaluates the condition based on the trigger and the event.
   * @param trigger the trigger
   * @param event the event
   * @return true if the condition is met, false otherwise
   */
  default boolean evaluateTriggerEvent(final @NotNull RegistrableTrigger trigger, @NotNull final Event event) {
    throw new UnsupportedOperationException("This has not been implemented.");
  }
}

File: src/main/kotlin/com/roughlyunderscore/registry/RegistrableCondition.kt
// Copyright 2024 RoughlyUnderscore
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.roughlyunderscore.registry

import com.roughlyunderscore.annotations.Since
import com.roughlyunderscore.enums.TargetType
import org.bukkit.event.Event

/**
 * This interface is used to create registrable conditions.
 */
@Since("2.2")
interface RegistrableCondition : Registrable {
  /**
   * Evaluates the condition, given the [trigger], [event], [target] and [arguments].
   * @return true if the condition is met, false otherwise
   * @since 2.2
   */
  fun evaluateCondition(trigger: RegistrableTrigger, event: Event, target: TargetType, arguments: List<String>): Boolean
}